# Toronto Neighborhood Exploration

## Part1 : Web Scraping from Wikipedia

### I used requests to get the html file form the wikipedia page and then tried to find any table contained in that html

In [91]:
import requests
import pandas as pd
from bs4 import BeautifulSoup
import numpy as np

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_page = requests.get(url)
soup = BeautifulSoup(wiki_page.content,'lxml')
table = soup.find_all('table')[0]
df = pd.read_html(str(table))[0]
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                                  NaN  
1                                                  NaN  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                                NaN  
176                                                NaN  
177                                                NaN  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                                NaN  

[180 rows x 3 columns]

### Drop all `Postal Code` that were `Not assigned`

In [71]:
df_drop = df[df.Borough != 'Not assigned']
df_drop

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [72]:
df_drop.shape

(103, 3)

## Part2 : Assign Latitudes and Longitudes

In [73]:
ll = pd.read_csv('Geospatial_Coordinates.csv')
ll.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [94]:
df_combine = df_drop.merge(ll,on='Postal Code')
df_combine.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## Part3 : Clustering Neighborhoods

### Display neighborhood on map with blue marker

In [97]:
import folium
latitude = 43.651070
longitude = -79.347015
toronto_map = folium.Map(location=[latitude,longitude],zoom_start=12)

In [76]:
for lat, lng, borough, neighborhood in zip(df_combine['Latitude'], df_combine['Longitude'], df_combine['Borough'], df_combine['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)
    
toronto_map

### Get nearby venues for each neighborhood (limit to 10 venues per area in the return list)

In [77]:
client_id = "WGENET1SWCZBUQYDFZDF5DGW5AIIE1THJCXUCQ1DW4PTHJNH"
client_secret = "CHCVAO54RWBGKU0IXZGZMVZCEG4DBCAPLOGNEWQNNW3V2MQJ"
version = "20180605"

In [78]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):           
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            client_id, 
            client_secret, 
            version, 
            lat, 
            lng, 
            radius, 
            10)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [79]:
venues = getNearbyVenues(names=df_combine['Neighborhood'],latitudes=df_combine.Latitude,longitudes=df_combine.Longitude)

### Venue list overview

In [80]:
venues.shape

(428, 7)

- In total `428` datapoints returned

In [81]:
venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1             649 Variety       43.754513       -79.331942  Convenience Store  
2           Variety Store       43.751974       -79.333114  Food & Drink Shop  
3  Victoria Village Arena       43.723481       -79.315635       Hockey Arena  
4             Tim Hortons       43.725517       -79.313103        Coffee Shop

### Which Neighborhood had most venues returned?

In [82]:
venues.groupby('Neighborhood').Venue.count().sort_values(ascending=False)

Neighborhood
Downsview                                        16
Don Mills                                         9
Leaside                                           5
Christie                                          5
Commerce Court, Victoria Hotel                    5
                                                 ..
The Kingsway, Montgomery Road, Old Mill North     2
Weston                                            2
York Mills, Silver Hills                          2
Rouge Hill, Port Union, Highland Creek            1
Humber Summit                                     1
Name: Venue, Length: 95, dtype: int64

- Seems like `Downsview` had 16 venues returned

### How many unique venue categories were in the dataset?

In [83]:
venues['Venue Category'].nunique()

150

In [84]:
venues['Venue Category'].value_counts()

Coffee Shop            28
Park                   23
Restaurant             16
Pizza Place            15
Café                   15
                       ..
Fried Chicken Joint     1
Health Food Store       1
Airport Terminal        1
Cuban Restaurant        1
Fish Market             1
Name: Venue Category, Length: 150, dtype: int64

- `150` unique venue categories
- `Coffee Shop` was appeared to be the most frequent venue category

### Most common venues for each neighborhood?

In [88]:
# one hot encoding
toronto_onehot = pd.get_dummies(venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = venues['Neighborhood']

toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Airport  \
0                                         Agincourt      0.0   
1                            Alderwood, Long Branch      0.0   
2   Bathurst Manor, Wilson Heights, Downsview North      0.0   
3                                   Bayview Village      0.0   
4                 Bedford Park, Lawrence Manor East      0.0   
..                                              ...      ...   
90                      Willowdale, Willowdale West      0.0   
91                                           Woburn      0.0   
92                                 Woodbine Heights      0.0   
93                                  York Mills West      0.0   
94                         York Mills, Silver Hills      0.0   

    Airport Food Court  Airport Lounge  Airport Terminal  American Restaurant  \
0                  0.0             0.0               0.0                  0.0   
1                  0.0             0.0               0.0                  0.0   
2                  0.0             0.0               0.0                  0.0   
3                  0.0             0.0               0.0                  0.0   
4                  0.0             0.0               0.0                  0.0   
..                 ...             ...               ...                  ...   
90                 0.0             0.0               0.0                  0.0   
91                 0.0             0.0               0.0                  0.0   
92                 0.0             0.0               0.0                  0.0   
93                 0.0             0.0               0.0                  0.0   
94                 0.0             0.0               0.0                  0.0   

    Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bakery  ...  \
0                   0.0               0.0                 0.0     0.0  ...   
1                   0.0               0.0                 0.0     0.0  ...   
2                   0.0               0.0                 0.0     0.0  ...   
3                   0.0               0.0                 0.0     0.0  ...   
4                   0.0               0.0                 0.0     0.0  ...   
..                  ...               ...                 ...     ...  ...   
90                  0.0               0.0                 0.0     0.0  ...   
91                  0.0               0.0                 0.0     0.0  ...   
92                  0.0               0.0                 0.0     0.0  ...   
93                  0.0               0.0                 0.0     0.0  ...   
94                  0.0               0.0                 0.0     0.0  ...   

    Theater  Theme Restaurant  Toy / Game Store  Trail  \
0       0.0               0.0               0.0    0.0   
1       0.0               0.0               0.0    0.0   
2       0.0               0.0               0.0    0.0   
3       0.0               0.0               0.0    0.0   
4       0.0               0.0               0.0    0.0   
..      ...               ...               ...    ...   
90      0.0               0.0               0.0    0.0   
91      0.0               0.0               0.0    0.0   
92      0.0               0.0               0.0    0.0   
93      0.0               0.0               0.0    0.0   
94      0.0               0.0               0.0    0.0   

    Vegetarian / Vegan Restaurant  Vietnamese Restaurant  Warehouse Store  \
0                             0.0                    0.0              0.0   
1                             0.0                    0.0              0.0   
2                             0.0                    0.0              0.0   
3                             0.0                    0.0              0.0   
4                             0.0                    0.0              0.0   
..                            ...                    ...              ...   
90                            0.0                    0.0              0.0   
91                            0.0              

In [89]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [92]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East       Thai Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1                   Gym                Coffee Shop                   Pub   
2         Deli / Bodega                Bridal Shop            Restaurant   
3                  Café        Japanese Restaurant    Chinese Restaurant   
4                  Café         Italian Restaurant     Indian Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0               Dog Run        Farmers Market            Electronics Store   
1          Dessert Shop             Drugstore                   Donut Shop   
2           Yoga Studio   Distribution Center            Electronics Store   
3  Fast Food Restaurant     Electronics Store  Eastern European Restaurant   
4            Restaurant        Discount Store  Eastern European Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant             Drugstore             Donut Shop  
1                      Dog Run   Distribution Center         Discount Store  
2  Eastern European Restaurant             Drugstore             Donut Shop  
3                    Drugstore            Donut Shop                Dog Run  
4                    Drugstore            Donut Shop                Dog Run

### Running k-means algorithm to cluster neighborhoods

In [93]:
from sklearn.cluster import KMeans
k = 5
toronto_cluster = toronto_grouped.drop('Neighborhood',axis=1)

model = KMeans(n_clusters=k,random_state=0)
model.fit(toronto_cluster)
model.labels_[0:10]

array([2, 1, 4, 3, 3, 3, 3, 4, 2, 2])

In [96]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', model.labels_)

toronto_merged = df_combine
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'),on='Neighborhood')
toronto_merged

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
..                                                 ...        ...        ...   
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   
99                                Church and Wellesley  43.665860 -79.383160   
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509   
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999   

     Cluster Labels       1st Most Common Venue  2nd Most Common Venue  \
0               0.0           Convenience Store                   Park   
1               1.0                 Pizza Place           Hockey Arena   
2               4.0              Breakfast Spot            Coffee Shop   
3               4.0                    Boutique  Vietnamese Restaurant   
4               4.0                    Creperie     Italian Restaurant   
..              ...                         ...                    ...   
98              0.0                       River                   Park   
99              2.0             Bubble Tea Shop           Dance Studio   
100             2.0                     Brewery         Farmers Market   
101             2.0  Construction & Landscaping         Baseball Field   
102             2.0                      Bakery         Discount Store   

    3rd Most Common Venue        4th Most Common Venue  \
0       Food & Drink Shop               Discount Store   
1             Coffee Shop        Portuguese Restaurant   
2                  Bakery                          Spa   
3             Coffee Shop               Clothing Store   
4                    Park                  Coffee Shop   
..                    ...                          ...   
98            Yoga Studio  Eastern European Restaurant   
99              Gastropub             Theme Restaurant   
100            Comic Shop                Burrito Place   
101           Yoga Studio               Discount Store   
102           Wings Joint                 Burger Joint   

      5th Most Common Venue        6th Most Common Venue  \
0         Electronics Store  Eastern European Restaurant   
1         French Restaurant                  Curling Ice   
2       Distribution Center                  Yoga Studio   
3    Furniture / Home Store                  Yoga Studio   
4       Distribution Center  Eastern European Restaurant   
..                      ...                          ...   
98                Drugstore                   Donut Shop   
99       Mexican Restaurant                  Yoga Studio   
100             Pizza Place        General Entertainment   
101       Electronics Store  Eastern European Restaurant   
102         Supplement Shop            Electronics Store   

           7th Most Common Venue        8th Most Common Venue  \
0                      Drugstore                   Donut Shop   
1               Cuban Restaurant                 Dance Studio   
2                        Dog Run               Farmers Market   
3            Distribution Cente

In [110]:
toronto_merged.dropna(subset=['Cluster Labels'],inplace=True)

In [113]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine each cluster

In [118]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0]

Postal Code           Borough  \
0          M3A        North York   
21         M6E              York   
35         M4J         East York   
61         M4N   Central Toronto   
64         M9N              York   
66         M2P        North York   
83         M4T   Central Toronto   
85         M1V       Scarborough   
91         M4W  Downtown Toronto   
98         M8X         Etobicoke   

                                         Neighborhood   Latitude  Longitude  \
0                                           Parkwoods  43.753259 -79.329656   
21                                Caledonia-Fairbanks  43.689026 -79.453512   
35      East Toronto, Broadview North (Old East York)  43.685347 -79.338106   
61                                      Lawrence Park  43.728020 -79.388790   
64                                             Weston  43.706876 -79.518188   
66                                    York Mills West  43.752758 -79.400049   
83                        Moore Park, Summerhill East  43.689574 -79.383160   
85  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
91                                           Rosedale  43.679563 -79.377529   
98      The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0              0.0     Convenience Store                  Park   
21             0.0                  Park                   Spa   
35             0.0     Convenience Store                  Park   
61             0.0                  Park              Bus Line   
64             0.0     Convenience Store                  Park   
66             0.0                  Park     Convenience Store   
83             0.0          Tennis Court            Playground   
85             0.0            Playground                  Park   
91             0.0                  Park            Playground   
98             0.0                 River                  Park   

   3rd Most Common Venue        4th Most Common Venue  \
0      Food & Drink Shop               Discount Store   
21         Women's Store               Discount Store   
35           Coffee Shop               Discount Store   
61           Swim School                  Yoga Studio   
64        Discount Store            Electronics Store   
66                  Bank            Electronics Store   
83                  Park                  Summer Camp   
85                 Diner  Eastern European Restaurant   
91                 Trail                  Yoga Studio   
98           Yoga Studio  Eastern European Restaurant   

          5th Most Common Venue        6th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21            Electronics Store  Eastern European Restaurant   
35            Electronics Store  Eastern European Restaurant   
61               Discount Store  Eastern European Restaurant   
64  Eastern European Restaurant                    Drugstore   
66  Eastern European Restaurant                    Drugstore   
83                  Yoga Studio                    Drugstore   
85                    Drugstore                   Donut Shop   
91                        Diner  Eastern European Restaurant   
98                    Drugstore                   Donut Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0              Drugstore            Donut Shop               Dog Run   
21             Drugstore            Donut Shop               Dog Run   
35             Drugstore            Donut Shop               Dog Run   
61             Drugstore            Donut Shop               Dog Run   
64            Donut Shop               Dog Run   Distribution Center   
66            Donut Shop               Dog Run   Distribution Center   
83            Donut Shop               Dog Run   Distribution Center   
85               Dog Run   Distribution Center        Discount Store   
91             Drugstore    

- Cluster 0 grouped neighborhoods with common venues like `Park`,`Convenience Store`

In [119]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1]

Postal Code      Borough  \
1          M4A   North York   
8          M4B    East York   
17         M9C    Etobicoke   
50         M9L   North York   
56         M6M         York   
70         M9P    Etobicoke   
72         M2R   North York   
76         M7R  Mississauga   
77         M9R    Etobicoke   
93         M8W    Etobicoke   

                                         Neighborhood   Latitude  Longitude  \
1                                    Victoria Village  43.725882 -79.315572   
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201   
50                                      Humber Summit  43.756303 -79.565963   
56   Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.691116 -79.476013   
70                                          Westmount  43.696319 -79.532242   
72                        Willowdale, Willowdale West  43.782736 -79.442259   
76              Canada Post Gateway Processing Centre  43.636966 -79.615819   
77  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724   
93                             Alderwood, Long Branch  43.602414 -79.543484   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1              1.0           Pizza Place          Hockey Arena   
8              1.0           Pizza Place                  Bank   
17             1.0           Pizza Place        Cosmetics Shop   
50             1.0           Pizza Place        Discount Store   
56             1.0  Fast Food Restaurant           Coffee Shop   
70             1.0           Pizza Place           Coffee Shop   
72             1.0           Pizza Place          Home Service   
76             1.0                 Hotel         Burrito Place   
77             1.0           Pizza Place     Mobile Phone Shop   
93             1.0           Pizza Place                   Gym   

   3rd Most Common Venue  4th Most Common Venue        5th Most Common Venue  \
1            Coffee Shop  Portuguese Restaurant            French Restaurant   
8              Gastropub               Pharmacy         Gym / Fitness Center   
17          Liquor Store             Beer Store                  Coffee Shop   
50        Farmers Market      Electronics Store  Eastern European Restaurant   
56        Sandwich Place         Discount Store            Electronics Store   
70        Sandwich Place     Chinese Restaurant                  Curling Ice   
72              Pharmacy            Coffee Shop                         Bank   
76           Coffee Shop         Sandwich Place                  Yoga Studio   
77                  Park         Sandwich Place                    Drugstore   
93           Coffee Shop                    Pub                 Dessert Shop   

          6th Most Common Venue        7th Most Common Venue  \
1                   Curling Ice             Cuban Restaurant   
8           American Restaurant          Distribution Center   
17               Discount Store  Eastern European Restaurant   
50                    Drugstore                   Donut Shop   
56  Eastern European Restaurant                    Drugstore   
70                 Dance Studio             Cuban Restaurant   
72                 Dance Studio                  Curling Ice   
76               Discount Store  Eastern European Restaurant   
77                   Donut Shop                      Dog Run   
93                    Drugstore                   Donut Shop   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
1           Dance Studio              Creperie                Deli / Bodega  
8         Farmers Market     Electronics Store  Eastern European Restaurant  
17             Drugstore            Donut Shop                      Dog Run  
50               Dog Run   Distribution Center                  Yoga Studio  
56            Donut Shop               Dog Run          Distribution Center  
70        

- Cluster 1 grouped neighborhoods with common venues like `Pizza Place`

In [120]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2]

Postal Code           Borough  \
6           M1B       Scarborough   
9           M5B  Downtown Toronto   
10          M6B        North York   
12          M1C       Scarborough   
14          M4C         East York   
16          M6C              York   
18          M1E       Scarborough   
19          M4E      East Toronto   
26          M1H       Scarborough   
27          M2H        North York   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
40          M3K        North York   
41          M4K      East Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
49          M6L        North York   
53          M3M        North York   
54          M4M      East Toronto   
57          M9M        North York   
59          M2N        North York   
60          M3N        North York   
62          M5N   Central Toronto   
63          M6N              York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
71          M1R       Scarborough   
75          M6R      West Toronto   
78          M1S       Scarborough   
84          M5T  Downtown Toronto   
87          M5V  Downtown Toronto   
89          M9V         Etobicoke   
90          M1W       Scarborough   
94          M9W         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
6                                       Malvern, Rouge  43.806686 -79.194353   
9                             Garden District, Ryerson  43.657162 -79.378937   
10                                           Glencairn  43.709577 -79.445073   
12              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
14                                    Woodbine Heights  43.695344 -79.318389   
16                                  Humewood-Cedarvale  43.693781 -79.428191   
18                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
19                                         The Beaches  43.676357 -79.293031   
26                                           Cedarbrae  43.773136 -79.239476   
27                                   Hillcrest Village  43.803762 -79.363452   
31                        Dufferin, Dovercourt Village  43.669005 -79.442259   
32                                 Scarborough Village  43.744734 -79.239476   
33                        Fairview, Henry Farm, Oriole  43.778517 -79.346556   
34                     Northwood Park, York University  43.767980 -79.487262   
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
37                            Little Portugal, Trinity  43.647927 -79.419750   
40                                           Downsview  43.737473 -79.464763   
41                        The Danforth West, Riverdale  43.679557 -79.352188   
44                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
45                            York Mills, Silver Hills  43.757490 -79.374714   
46                                           Downsview  43.739015 -79.506944   
47                      India Bazaar, The Beaches West  43.668999 -79.315572   
49            North Park, Maple Leaf Park, Upwood Park  43.713756 -79.490074   
53                                           Downsview  43.728496 -79.495697   
54                                     Studio District  43.659526 -79.340923   
57                                    Humberlea, Emery  43.724766 -79.532242   
59                         Willowdale, Willowdale East  43.770120 -79.408493   
60                                           Downsview  43.76

- Cluster 2 grouped neighborhoods with common venues like `Donut Shop`,`Drugstore`,`Dog Run`

In [121]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3]

Postal Code           Borough  \
7          M3B        North York   
13         M3C        North York   
20         M5E  Downtown Toronto   
23         M4G         East York   
29         M4H         East York   
30         M5H  Downtown Toronto   
39         M2K        North York   
48         M5L  Downtown Toronto   
51         M1M       Scarborough   
55         M5M        North York   
58         M1N       Scarborough   
65         M1P       Scarborough   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
82         M1T       Scarborough   
86         M4V   Central Toronto   
88         M8V         Etobicoke   
92         M5W  Downtown Toronto   
96         M4X  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
7                                           Don Mills  43.745906 -79.352188   
13                                          Don Mills  43.725900 -79.340923   
20                                        Berczy Park  43.644771 -79.373306   
23                                            Leaside  43.709060 -79.363452   
29                                   Thorncliffe Park  43.705369 -79.349372   
30                           Richmond, Adelaide, King  43.650571 -79.384568   
39                                    Bayview Village  43.786947 -79.385975   
48                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
51    Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
55                  Bedford Park, Lawrence Manor East  43.733283 -79.419750   
58                        Birch Cliff, Cliffside West  43.692657 -79.264848   
65  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
73                  North Toronto West, Lawrence Park  43.715383 -79.405678   
74                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
79                                         Davisville  43.704324 -79.388790   
80                     University of Toronto, Harbord  43.662696 -79.400049   
82            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
86  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
88       New Toronto, Mimico South, Humber Bay Shores  43.605647 -79.501321   
92                                     Stn A PO Boxes  43.646435 -79.374846   
96                        St. James Town, Cabbagetown  43.667967 -79.367675   

    Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
7              3.0                   Gym             Restaurant   
13             3.0                   Gym             Restaurant   
20             3.0          Liquor Store           Concert Hall   
23             3.0   Sporting Goods Shop             Restaurant   
29             3.0           Yoga Studio      Indian Restaurant   
30             3.0          Concert Hall                  Plaza   
39             3.0                  Bank                   Café   
48             3.0                   Gym             Restaurant   
51             3.0         Movie Theater    American Restaurant   
55             3.0       Thai Restaurant                   Café   
58             3.0          Skating Rink  General Entertainment   
65             3.0     Indian Restaurant  Vietnamese Restaurant   
73             3.0           Yoga Studio             Restaurant   
74             3.0                  Park    American Restaurant   
79             3.0          Dessert Shop            Pizza Place   
80             3.0            Restaurant           Dessert Shop   
82             3.0       Thai Restaurant           Noodle House   
86             3.0          Liquor Store             Restaurant   
88             3.0           Pizza Place                 Bakery   
92             3.0              Fountain             Restaurant   
96             3.0            Restaurant    Japanese Restaurant 

- Cluster 3 grouped neighborhoods with common venues like `Gym`,`Cafe`,`Restaurant`

In [122]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4]

Postal Code           Borough  \
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
15         M5C  Downtown Toronto   
22         M1G       Scarborough   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
28         M3H        North York   
38         M1K       Scarborough   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
81         M6S      West Toronto   
97         M5X  Downtown Toronto   

                                       Neighborhood   Latitude  Longitude  \
2                         Regent Park, Harbourfront  43.654260 -79.360636   
3                  Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4       Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
15                                   St. James Town  43.651494 -79.375418   
22                                           Woburn  43.770992 -79.216917   
24                               Central Bay Street  43.657952 -79.387383   
25                                         Christie  43.669542 -79.422564   
28  Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
38      Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
42         Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
43     Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
81                               Runnymede, Swansea  43.651571 -79.484450   
97           First Canadian Place, Underground city  43.648429 -79.382280   

    Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
2              4.0        Breakfast Spot            Coffee Shop   
3              4.0              Boutique  Vietnamese Restaurant   
4              4.0              Creperie     Italian Restaurant   
15             4.0           Coffee Shop               Creperie   
22             4.0           Coffee Shop      Korean Restaurant   
24             4.0           Coffee Shop                    Spa   
25             4.0                  Café     Italian Restaurant   
28             4.0           Coffee Shop          Deli / Bodega   
38             4.0        Discount Store     Chinese Restaurant   
42             4.0                   Gym             Restaurant   
43             4.0           Coffee Shop                    Gym   
81             4.0                  Café      Fish & Chips Shop   
97             4.0                   Gym   Gym / Fitness Center   

        3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
2                      Bakery                   Spa     Distribution Center   
3                 Coffee Shop        Clothing Store  Furniture / Home Store   
4                        Park           Coffee Shop     Distribution Center   
15        Japanese Restaurant            Restaurant            Dance Studio   
22                Yoga Studio        Discount Store       Electronics Store   
24  Middle Eastern Restaurant           Yoga Studio          Discount Store   
25              Grocery Store           Coffee Shop             Yoga Studio   
28                Bridal Shop            Restaurant             Yoga Studio   
38                Coffee Shop      Department Store       Electronics Store   
42                Coffee Shop                  Café                  Bakery   
43                        Bar    Italian Restaurant          Farmers Market   
81                Coffee Shop                  Food        Sushi Restaurant   
97                Coffee Shop                Bakery              Restaurant   

          6th Most Common Venue        7th Most Common Venue  \
2                   Yoga Studio                      Dog Run   
3                   Yoga Studio          Distribution Center   
4   Eastern European Restaurant                    Drugstore   
15                  Curling Ice                Deli / Bodega   
22  Eastern European Restaurant                    Dr

- Cluster 4 grouped neighborhoods with common venues like `Coffee Shop`